keyboards/snes_macropad/snes_macropad.c

```diff
@@ -0,0 +1,130 @@
+    } else if (IS_QK_MODS(keycode)) {
+        keycode = QK_MODS_GET_BASIC_KEYCODE(keycode);
+    } else if (IS_QK_ONE_SHOT_MOD(keycode)) {
+        keycode = 0xE0 + biton(QK_ONE_SHOT_MOD_GET_MODS(keycode) & 0xF) + biton(QK_ONE_SHOT_MOD_GET_MODS(keycode) & 0x10);
+    }
+    if (keycode > ARRAY_SIZE(code_to_name)) {
+        return;
+    }
+
+    // update keylog
+    key_name = pgm_read_byte(&code_to_name[keycode]);
+}
+
+static const char *depad_str(const char *depad_str, char depad_char) {
+    while (*depad_str == depad_char) {
+        ++depad_str;
+    }
+    return depad_str;
+}
+
+static void oled_render_keylog(void) {
+    oled_write_char('0' + last_row, false);
+    oled_write("x", false);
+    oled_write_char('0' + last_col, false);
+    oled_write(", k", false);
+    const char *last_keycode_str = get_u16_str(last_keycode, ' ');
+    oled_write(depad_str(last_keycode_str, ' '), false);
+    oled_write(":", false);
+    oled_write_char(key_name, false);
+}
+
+__attribute__((weak)) const char * get_layer_name_user(int layer) {
+    return "Unknown";
+}
+
+static void oled_render_layer(void) {
+    oled_write("Layer: ", false);
+    oled_write_ln(get_layer_name_user(get_highest_layer(layer_state)), false);
+}
+
+bool oled_task_kb(void) {
+    if (!oled_task_user()) {
+        return false;
+    }
+
+    oled_render_layer();
+    oled_render_keylog();
+    oled_advance_page(true);
+    return false;
+}
+
+static void setupForFlashing(void) {
+    oled_clear();
+    oled_write("                     ", false);
+    oled_write("  In flash mode...   ", false);
+    oled_write("                     ", false);
+    oled_write("                     ", false);
+
+    // QMK is clever about only rendering a certain number of chunks per frame,
+    // but since the device will go into flash mode right after this call,
+    // we want to override this behavior and force all the chunks to be sent to
+    // the display immediately.
+    const size_t numIterations = OLED_DISPLAY_WIDTH * OLED_DISPLAY_HEIGHT / OLED_UPDATE_PROCESS_LIMIT;
+    for (size_t num = 0; num < numIterations; ++num) {
+        oled_render();
+    }
+    // todo: Replace the above hack with this, once develop branch is merged at the end of November 2023
+    // oled_render_dirty(true);
+
+    // Set alternating backlight colors
+    const uint8_t max = 20;
+    rgblight_mode_noeeprom(RGBLIGHT_MODE_STATIC_LIGHT);
+    for (size_t i = 0; i < RGBLED_NUM; ++i) {
+        LED_TYPE *led_ = (LED_TYPE *)&led[i];
+        switch (i % 2) {
+            case 0:
+                setrgb(max, 0, max, led_);
+                break;
+            case 1:
+                setrgb(0, max, max, led_);
+                break;
+        }
+    }
+    rgblight_set();
+}
+
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+    if (record->event.pressed) {
+        set_keylog(keycode, record);
+    }
+    if (keycode == QK_BOOT) {
+        setupForFlashing();
```

Comment:
Another thing to note here, is that `shutdown_user` is called by `QK_BOOT`.   Though, the caveats are that it currently has no way to differentiate between rebooting and jumping to bootloader, without adding manual tracking.  Also, you'd need to use `__attribute__((weak)) void shutdown_user(void)` since there is no kb version, currently.

Reply:
Sounds like the current implementation is simpler, or do you think I should change it? 

Either way I'm thinking it might make sense to mark it weak, to allow keymaps to override its behavior. But I'm currently putting that in the future improvements bucket.

Reply:
For now, this implementation is fine.  If you wanted to change things when rebooting, not just jumping to bootloader, then shutdown_user would be better.  But this works just fine.